rpcserver.go

```diff
@@ -3324,18 +3324,33 @@ func handleSendRawTransaction(s *rpcServer, cmd interface{}, closeChan <-chan st
 		// When the error is a rule error, it means the transaction was
 		// simply rejected as opposed to something actually going wrong,
 		// so log it as such.  Otherwise, something really did go wrong,
-		// so log it as an actual error.  In both cases, a JSON-RPC
-		// error is returned to the client with the deserialization
-		// error code (to match bitcoind behavior).
-		if _, ok := err.(mempool.RuleError); ok {
+		// so log it as an actual error.
+		ruleErr, ok := err.(mempool.RuleError)
+		if ok {
 			rpcsLog.Debugf("Rejected transaction %v: %v", tx.Hash(),
 				err)
 		} else {
 			rpcsLog.Errorf("Failed to process transaction %v: %v",
 				tx.Hash(), err)
 		}
+
+		// We'll then map the rule error to the appropriate RPC error,
+		// matching bitcoind's behavior.
+		code := btcjson.ErrRPCTxError
+		if txRuleErr, ok := ruleErr.Err.(mempool.TxRuleError); ok {
+			errDesc := txRuleErr.Description
+			switch {
+			case strings.Contains(errDesc, "orphan transaction"):
+				code = btcjson.ErrRPCTxError
+			case strings.Contains(errDesc, "transaction already exists"):
+				code = btcjson.ErrRPCTxAlreadyInChain
+			default:
+				code = btcjson.ErrRPCTxRejected
+			}
+		}
+
 		return nil, &btcjson.RPCError{
-			Code:    btcjson.ErrRPCDeserialization,
```

Comment:
🙈

Reply:
Ah guess it was OK in the past since we directly parsed the error string itself. Fixing this now will let us rely solely on the error code in the future though, which is nice.